WIP: Show tasks and allocations for instances

swsnr opened the pull request:
Corresponding project: https://github.com/FRosner/cluster-broccoli/projects/2

swsnr committed:
Add basic HTTP client to get allocations for jobs

swsnr committed:
GH-44: Add endpoint to get allocations for instances

Without tests unfortunately because InstanceService carries too much
state to create instances of it in tests.

swsnr committed:
Run integration tests sequentially

swsnr committed:
Test against a dockerized controlled nomad instance

swsnr committed:
Run SBT integration tests on Travis CI

FRosner commented on server/src/main/scala/de/frosner/broccoli/nomad/NomadConfiguration.scala:
```diff
@@ -0,0 +1,21 @@
+  *
+  * @param url The URL to connect to to access nomad via HTTP.
+  */
+final case class NomadConfiguration(url: String)
```
What happens here if the url is not set?

swsnr replied:
In application.conf?  It'll fallback to the one from reference.conf.

FRosner replied:
What will happen to the application if neither application.conf nor reference.conf contains this value?

swsnr replied:
It'll throw an exception.  reference.conf is supposed to contain all configuration settings you use throughout the application; hence the name.

swsnr replied:
PS: reference.conf is not for the user of an application: You shouldn't modify reference.conf during deployment; it's part of the application code essentially

FRosner commented:
Is the reference.conf packaged inside the jar or the dist zip automatically?

Does play exit cleanly if there is an exception at this point? Will it respect shutdown hooks?

Thanks!

swsnr commented:
@FRosner  Why should `reference.conf` not be packaged?  It's a resource like any other, isn't it?

I think that Play runs stop hooks no matter how the application exits, but I would not bother to find out in detail.  

As I said `reference.conf` should contain **all** settings of the application; if an exception gets thrown here, that's a programmer error.

swsnr committed:
Merge remote-tracking branch 'origin/master' into swiesner/issue-44-show-tasks-and-allocations-for-instances

* origin/master:
  Rename elm directory to webui
  Remove shell scripts in favour of direct sbt dist
  Build frontend with SBT
  #229 Use ./templates for the templates path in  reference.conf
  #229 Remove Singleton annotations from SignalRefreshedTemplateSource and CachedTemplateSource
  #229 Reword docs for TemplateSource and CachedTemplateSource
  #229 Trim spaces in a template before computing version
  #229 Move templates into /cluster-broccoli-dist in the test docker image
  #229 Use absolute paths when loading templates from a directory
  #229 Add documentation to Meta case class
  #229 Inject TemplateModule into TemplateService
  #229 Refresh templates by sending SIGUSR2
  #229 Refactor TemplateService.templates
  Enable hot module reloading for Elm
  Redo frontend build

swsnr committed:
Update SBT

swsnr committed:
Define task model

swsnr committed:
Extract instance tasks

swsnr committed:
Strengthen type of webservice messages

Replace any-typed payload with sealed trait enum, and move from iterable to Seq

swsnr committed:
Fix compile task in package.json

Actually compile the main module :)

swsnr committed:
Send instance tasks over web sockets

swsnr committed:
Use a type-safe representation of websocket messages

Instead of attaching types to JSON blobs, define messages as ADTs with explicit
connection between the type and the message.

Decode as early and encode as late as possible, and otherwise use the typesafe
representations of messages.

Greatly reduce possible error cases and simplify code.

swsnr committed:
Replace MaybeUtils with maybe-extra

swsnr committed:
Show tasks and allocs when instance is expanded

swsnr commented:
@FRosner This is how it looks like with the latest push:

<img width="608" alt="bildschirmfoto 2017-08-02 um 17 28 23" src="https://user-images.githubusercontent.com/224922/28881433-62736c32-77a8-11e7-8c16-29f9cf2e6f41.PNG">

swsnr committed:
Merge remote-tracking branch 'origin/master' into swiesner/issue-44-show-tasks-and-allocations-for-instances

* origin/master:
  Mock SignalManager in SignalRefreshedTemplateSource tests
  Annotate UnixSignalManager as a singleton
  Use configure method instead of provideSignalManager in SignalModule
  Improve docstrings in SignalManager
  Enable SignalModule
  Extract triggering unix signals into integration tests
  Document exceptions that SignalManager can throw
  Do not throw an exception when unregistering a signal that was not registered
  Explicitly test signal registration with distinct signal objects
  Refactor signal handling into separate module

swsnr committed:
Refresh instance tasks of expanded instances

swsnr committed:
Read task states as nullable

swsnr committed:
Filter complete allocations to avoid confusing the user

swsnr committed:
Check access to instance when retrieving instance logs

swsnr committed:
Apply code suggestions

Remove an unused name and simplify option value check

swsnr committed:
Communicate errors for get instance tasks over websocket

Introduce a generic "InstanceError" model to represent instance errors

The instance service returns an Either of error or tasks.  The controllers
convert the Either into a websocket message or an HTTP result.

Introduce a ToHttpResult typeclass to abstract over the latter.

swsnr committed:
Mark specialized Result types as deprecated

At some point we should convert to Either over InstanceError to
represent to outcome of instance operations

swsnr commented:
@FRosner I'll close this PR and split it into smaller ones for easier review.
